### DISCUSSION TITLE
Accessing Ray Differentials in BSDF's sample() function

### DISCUSSION THREAD

**Discussion by AndrewYiBC** [mitsuba-renderer/mitsuba3#1209] [Q&A] [answered]
Greetings,

I am creating a custom BSDF which requires ray differentials to determine the ray's corresponding pixel footprint in texture space. However, currently I do not see a way of accessing the ray differentials' (or even the ray's) information in the BSDF's sample() function.

Therefore, may I know if it is possible to do so with a basic BSDF class? If not, may I have some suggestions on how to achieve it, possibly by implementing a custom integrator / renderer?

Any insights would be greatly appreciated!

**Comment by njroussel** [comment, marked as answer]
Hi @AndrewYiBC 

There's a chain of methods to handle this, when fetching the `bsdf` of an interaction you should use:  [`SurfaceInteraction3f.bsdf()`](https://mitsuba.readthedocs.io/en/latest/src/api_reference.html#mitsuba.SurfaceInteraction3f.bsdf). That method will call [`SurfaceInteraction3f.compute_uv_partials()`](https://mitsuba.readthedocs.io/en/latest/src/api_reference.html#mitsuba.SurfaceInteraction3f.compute_uv_partials) which uses the ray's partials (`ray.o_x` and `ray.o_y`). Finally, that last method call will actually fill in the `duv_dx`, `duv_dy` fields in your `si` object. This only works well for the first bounce, after which the ray partials don't correctly account for the warping introduced by the first bounce's BSDF.

**Reply by AndrewYiBC** [reply]
Hello @njroussel ,

Thank you so much for your reply. I tried calling `si.bsdf()` but it doesn't seem to fill the `duv_dx` and `duv_dy` field of my `si` object (they remain as `[0.0, 0.0]`). I am also unable to call `si.compute_uv_partials()` on my own as it requires the parameter `ray (mitsuba.core.RayDifferential3f)` which I don't have access to in `bsdf`'s `sample()` function (same for calling `si.bsdf(ray)`). Are there any ways to solve this issue or to access the ray reference?

Your assistance is truly appreciated!

**Reply by ziyi-zhang** [reply]
Hi, 
You need to call `si.bsdf(ray)`. The one without `ray` argument simply returns the BSDF without `compute_uv_partials()`.

Overall, differentials are not well handled in Mitsuba3 (whose focus is mainly on inverse rendering). E.g., a common use case of differentials is to select a Mipmap level, but a mipmap is still not supported in Mitsuba3. It would be easier to write your own code for such features using Mitsuba 0.6 / PBRT as reference.

Best

**Reply by AndrewYiBC** [reply]
Hello @ziyi-zhang ,

Thank you for the clarifications and suggestions. However, I couldn't seem to find a way to reference `ray` in `bsdf`'s `sample()` function, thus I am unable to use it as the argument for `si.bsdf(ray)`. May I know how this could be resolved? Much appreciated!

**Reply by njroussel** [reply]
This is usually done at the intergrator level, not inside the BSDF.

Have a look at our pathtracer to see how this is usually done: https://github.com/mitsuba-renderer/mitsuba3/blob/master/src/integrators/path.cpp#L203

**Reply by AndrewYiBC** [reply]
Hello @njroussel ,

Many thanks for your time and clarifications! I will take a close look at the integrator implementation.
